File: e2e/integration_tests/keras_to_tfjs_create_save_predict_test.py
# Copyright 2023 Google LLC
#
# Use of this source code is governed by an MIT-style
# license that can be found in the LICENSE file or at
# https://opensource.org/licenses/MIT.
# =============================================================================

# This file is 1/3 of the test suites for keras v3: create->save->predict.
# E2E from Keras model -> TFJS model -> Comparsion between two models.
#
# This file does below things:
# - Build Keras v3 model
# - Save the model to local files.
# - Convert and save the model to TFJS layers format.
# - Extract data and shape from model and store in local files.
import argparse
import json
import numpy as np
import tensorflow as tf
import tempfile
import os
from tensorflow import keras
from tensorflowjs.converters.converter import dispatch_keras_keras_to_tfjs_layers_model_conversion

print('tf-version: ', tf.version.VERSION)
curr_dir = os.path.dirname(os.path.realpath(__file__))
_tmp_dir = os.path.join(curr_dir, 'keras_to_tfjs_create_save_predict_data')

def _export_mlp_model(export_path, model_name):
  model = keras.Sequential()
  d = keras.layers.Dense(100, activation='relu')
  model.add(d)
  model.add(keras.layers.Dense(50, activation='elu'))
  model.add(keras.layers.Dense(24))
  model.add(keras.layers.Activation(activation='elu'))
  model.add(keras.layers.Dense(8, activation='softmax'))
  model.build(input_shape=[1, 200])

  save_model_and_random_inputs(model, export_path, model_name)

def save_model_and_random_inputs(model, export_path, model_name):
  os.mkdir(export_path)
  keras_file_path = export_path + '/' + model_name + '.keras'
  model.save(keras_file_path)
  print('export path: ', export_path)
  print('keras path: ', keras_file_path)
  dispatch_keras_keras_to_tfjs_layers_model_conversion(keras_file_path, output_dir=export_path)

  xs = []
  xsData = []
  xsShapes = []
  tensors = model.variables
  mInput = model.inputs
  print(mInput)

  for input_tensor in mInput:
    input_shape = input_tensor.shape
    print(input_shape)
    xTensor = tf.random.normal(input_shape)
    xs.append(xTensor)
    xsData.append(xTensor.numpy()[0])
    xsShapes.append(xTensor.shape)
  xs_data = [x.tolist() for x in xsData]
  xs_shape = [list(x) for x in xsShapes]

  xs_shape_path = os.path.join(_tmp_dir, model_name + '.xs-shapes.json')
  xs_data_path = os.path.join(_tmp_dir, model_name + '.xs-data.json')
  with open(xs_data_path, 'w') as f:
    f.write(json.dumps(xs_data))

  with open(xs_shape_path, 'w') as f:
    f.write(json.dumps(xs_shape))

  z = [np.array(value, dtype=np.float32).reshape(shape)
        for value, shape in zip(xs_data, xs_shape)]
  if len(z) == 1:
    z = z[0]

  k = model.predict(z)
  keras_predict_data = None
  keras_predict_shape = None

  if isinstance(k, list):
    keras_predict_data = [y.tolist() for y in k]
    keras_predict_shape = [list(y.shape) for y in k]
  else:
    keras_predict_data = k.tolist()
    keras_predict_shape = [list(k.shape)]

  result_data_path = os.path.join(
      _tmp_dir, model_name + '.result-from-keras-data.json')
  result_shape_path = os.path.join(
      _tmp_dir, model_name + '.result-from-keras-shape.json')
  with open(result_data_path, 'w') as f:
    f.write(json.dumps(keras_predict_data))
  with open(result_shape_path, 'w') as f:
    f.write(json.dumps(keras_predict_shape))


def main():
  print('tensorflow version: ', tf.__version__)

  parser = argparse.ArgumentParser(description='Create a keras model in python.')
  parser.add_argument('--test_data_dir', help='Input a directory name', default='test_data_dir')
  args = parser.parse_args()
  folder = args.test_data_dir
  _export_mlp_model(os.path.join(folder, 'mlp'), 'mlp')

if __name__ == '__main__':
  main()

File: e2e/integration_tests/keras_to_tfjs_create_save_predict_test.js
/**
 * @license
 * Copyright 2023 Google LLC. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * =============================================================================
 */

/**
 * This file is 2/3 of the test suites for keras v3: create->save->predict.
 * E2E from Keras model -> TFJS model -> Comparsion between two models.
 *
 * This file does below things:
 * - Load the TFJS layers model to runtime.
 * - Load inputs data.
 * - Make inference and store in local files.
 */
const tf = require('@tensorflow/tfjs');
const tfc = require('@tensorflow/tfjs-core');
const tfconverter = require('@tensorflow/tfjs-converter');
const tfl = require('@tensorflow/tfjs-layers');
const tfjsNode = require('@tensorflow/tfjs-node');
const ndarray = require('ndarray');
const fs = require('fs');
const join = require('path').join;
const TEST_DATA_DIR = "keras_to_tfjs_create_save_predict_data";

async function load_file(file_path, model_name) {

  const p = tfjsNode.io.fileSystem(file_path)

  const model = await tf.loadLayersModel("file://" + tfjsNode.io.fileSystem(file_path).path + "/model.json");

  const xsDataPath = join(TEST_DATA_DIR, `${model_name}.xs-data.json`);
  const xsShapesPath = join(TEST_DATA_DIR, `${model_name}.xs-shapes.json`);
  const dataArray = fs.readFileSync(xsDataPath, "utf-8", (_, data) => {
    return JSON.parse(data);
  });
  const shapeArray = fs.readFileSync(xsShapesPath, "utf-8", (_, data) => {
    return JSON.parse(data);
  });

  const parsedShape = JSON.parse(shapeArray);
  const parsedData = JSON.parse(dataArray);
  const zippedArray = zip(parsedData, parsedShape);
  console.log("zipped Array: ", zippedArray);
  const res = [];
  for (let i = 0; i < zippedArray.length; i++) {
    const value = zippedArray[i][0];
    const shape = zippedArray[i][1];
    res.push(ndarray(value, shape));
    // res.push(tfjsNode.tensor(value, shape));
  }
  let z;
  if (res.length == 1) {
    z = res[0];
  } else {
    z = res;
  }
  const tensor = tfjsNode.tensor(z.data, z.shape);
  const ys = model.predict(tensor);

  const resultData = ys.arraySync();
  const resultShape = [ys.shape];
  const ysDataPath = join(TEST_DATA_DIR, `${model_name}.ys-data.json`);
  const ysShapesPath = join(TEST_DATA_DIR, `${model_name}.ys-shapes.json`);
  fs.writeFileSync(ysDataPath, JSON.stringify(resultData));
  fs.writeFileSync(ysShapesPath, JSON.stringify(resultShape));
}

function zip(...arrays) {
  const length = Math.min(...arrays.map(arr => arr.length));
  const result = [];
  for (let i = 0; i < length; i++) {
    const zippedItem = arrays.map(arr => arr[i]);
    result.push(zippedItem);
  }
  return result;
}
console.log(`Using tfjs-core version: ${tfc.version_core}`);
console.log(`Using tfjs-layers version: ${tfl.version_layers}`);
console.log(`Using tfjs-node version: ${JSON.stringify(tfjsNode.version)}`);

(async function () {
  await load_file(join(TEST_DATA_DIR, "mlp"), "mlp");
})();
